File: Easy/problem20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                st.push(')');
            } else if (s[i] == '[') {
                st.push(']');
            } else if (s[i] == '{') {
                st.push('}');
            } else {
                if (st.empty()) {
                    return false;
                }
                char c = st.top(); 
                st.pop();
                if (s[i] != c) {
                    return false;
                }
            }
        }
        if (st.empty() == false) {
            return false;
        }
        return true;
    }
};

File: SolutionTypes/Stack/problem20.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        stack = []

        for char in s:
            if char == '(':
                stack.append(')')
            elif char == ')':
                if len(stack) == 0:
                    return False
                if stack[-1] != ')' :
                    return False
                stack.pop()
            elif char == '[':
                stack.append(']')
            elif char == ']':
                if len(stack) == 0:
                    return False
                if stack[-1] != ']' :
                    return False
                stack.pop()
            elif char == '{':
                stack.append('}')
            elif char == '}':
                if len(stack) == 0:
                    return False
                if stack[-1] != '}' :
                    return False
                stack.pop()
            
        if len(stack) > 0:
            return False

        return True
